espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/HamcrestMatcher.kt

```diff
@@ -0,0 +1,96 @@
+package io.appium.espressoserver.lib.model
+
+import com.google.gson.JsonDeserializationContext
+import com.google.gson.JsonDeserializer
+import com.google.gson.JsonElement
+import com.google.gson.JsonParseException
+import com.google.gson.annotations.JsonAdapter
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
+import io.appium.espressoserver.lib.helpers.GsonParserHelpers
+import io.appium.espressoserver.lib.helpers.KReflectionUtils
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import java.lang.reflect.Type
+import kotlin.reflect.KClass
+
+@JsonAdapter(HamcrestMatcher.HamcrestMatcherDeserializer::class)
+data class HamcrestMatcher (var name:String, var args:Array<Any?>, var matcherClass:KClass<*> = Matchers::class) {
+
+    fun invoke():Matcher<*> {
+        val matcher = KReflectionUtils.invokeMethod(this.matcherClass, this.name, *this.args)
+        if (matcher !is Matcher<*>) {
+            throw InvalidArgumentException("'${this}' does not return a Matcher when invoked. Found '${matcher!!::class.qualifiedName}'");
+        }
+        return matcher;
+    }
+
+    class HamcrestMatcherDeserializer : JsonDeserializer<HamcrestMatcher> {
+        @Throws(JsonParseException::class)
+        override fun deserialize(json: JsonElement, paramType: Type?,
+                                 paramJsonDeserializationContext: JsonDeserializationContext?): HamcrestMatcher {
+
+            if (json.isJsonObject) {
+
+                val jsonObj = json.asJsonObject
+
+                // Validate and parse the name property
+                val nameProp = jsonObj.get("name")
+
+                if (nameProp == null) {
+                    throw JsonParseException("Matcher must contain 'name' property")
+                } else if (!nameProp.isJsonPrimitive) {
+                    throw JsonParseException("'name' property on matcher must be a JSON primitive")
+                }
+
+                val name = nameProp.asString
+
+                // Parse args property
+                val listOfArgs = GsonParserHelpers.asArray(jsonObj, "args")
+
+                val args = arrayListOf<Any?>()
+                for (arg in listOfArgs) {
+                    if (arg.isJsonPrimitive) {
```

Comment:
[nits]
maybe, here also can write like:

```kotlin
when {
    arg.isJsonPrimitive -> args.add(GsonParserHelpers.parsePrimitive(arg.asJsonPrimitive))
    arg.isJsonNull -> args.add(null)
    arg.isJsonObject -> args.add(HamcrestMatcherDeserializer().deserialize(arg, null, null).invoke())
}
```

Reply:
Updated. I like that syntax.